staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -30,27 +31,53 @@ const (
 )
 
 // ready is a three state condition variable that blocks until is Ready if is not Stopped.
-// Its initial state is Pending.
+// Its initial state is Pending and its state machine diagram is as follow.
+//
+// Pending <------> Ready -----> Stopped
+//
+//	|                           ^
+//	└---------------------------┘
 type ready struct {
-	state status
-	c     *sync.Cond
+	state       status        // represent the state of the variable
+	lock        sync.RWMutex  // protect the state variable
+	restartLock sync.Mutex    // protect the transition from ready to pending where the channel is recreated
+	waitCh      chan struct{} // blocks until is ready or stopped
 }
 
 func newReady() *ready {
 	return &ready{
-		c:     sync.NewCond(&sync.RWMutex{}),
-		state: Pending,
+		waitCh: make(chan struct{}),
+		state:  Pending,
 	}
 }
 
+// done close the channel once the state is Ready or Stopped
+func (r *ready) done() chan struct{} {
+	r.restartLock.Lock()
+	defer r.restartLock.Unlock()
+	return r.waitCh
+}
+
 // wait blocks until it is Ready or Stopped, it returns an error if is Stopped.
-func (r *ready) wait() error {
-	r.c.L.Lock()
-	defer r.c.L.Unlock()
-	for r.state == Pending {
-		r.c.Wait()
+func (r *ready) wait(ctx context.Context) error {
+	// r.done() only blocks if state is Pending
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.done():
 	}
+
+	r.lock.RLock()
+	defer r.lock.RUnlock()
 	switch r.state {
+	case Pending:
+		// since we allow to switch between the states Pending and Ready
+		// if there is a quick transition from Pending -> Ready -> Pending
+		// a process that was waiting can get unblocked and see a Pending state.
+		// If the state is Pending don't return  an error because it can only happen
+		// here after the r.done() channel is closed because the state moved from
+		// Pending to Ready.
```

Comment:
I think it's not related to this PR, but I'm looking from watchcache code and I think we may have a bug which I want to callout here, so we make a conscious decision what to do with this PR in the meantime (or let me know if I'm wrong of course).

It's basically a cacher error, not ready error - the scenario is the following:

0. `ListAndWatch` in `startCaching` fails (for whatever reason)
1. new watch is being established `Watch()` method
2. ready is initialized so we successfully pass the `c.ready.wait()` check
3. `c.ready.set(false)` is called (from defer in `startCaching`)
4. `startCaching` is called again and we call `terminateAllWatchers()`
5. the Watch call (started in point 1 above) continues to be processed and is established (and added to watchers) - init events are being send in the meantime
6. `ListAndWatch` is being called from `startCaching` which reinitializes watchCache to a new (higher RV)
7. if some events missed in the meantime, they will not be delivered to this newly established watch


Now - I think it's highly unlikely now, because `startCaching` is called using `Until` with 1s period and it's fairly unlikely that Watch() will be hanging for a whole second between after calling ready.wait() and finishing.
But if I'm right, this seems like a serious correctness flaw.

@aojea - can you please double check my reasoning above?
@liggitt - FYI ^^

Reply:
yeah, is the same problem as the one we have here https://github.com/kubernetes/kubernetes/pull/116024#discussion_r1118366920 but within the cacher ... we have that window open :)

The logic of the cacher is too complex to me to fully understand it, why `terminateAllWatchers()` doesn't or can't reset the running watch, or just fully terminate it?

Reply:
OK - I'm fairly sure this PR is not making this problem worse/better.
I opened https://github.com/kubernetes/kubernetes/issues/116122 for it - will track it there.